File: packages/python/src/sonicwall/resources/_normalize.py
"""Shared response normalization helpers for firmware-variant envelopes."""

from __future__ import annotations

from typing import Any, Callable


def normalize_get_from_plural(
    response: dict[str, Any],
    *,
    plural_key: str,
    singular_key: str,
    predicate: Callable[[dict[str, Any]], bool] | None = None,
) -> dict[str, Any]:
    """Normalize list-envelope get response into a singular envelope.

    Example:
      {"address_objects": [{"ipv4": {...}}]} -> {"address_object": {"ipv4": {...}}}
    """
    items = response.get(plural_key)
    if not isinstance(items, list) or not items:
        return response

    selected: dict[str, Any] | None = None
    if predicate is not None:
        for item in items:
            if isinstance(item, dict) and predicate(item):
                selected = item
                break
    if selected is None:
        first = items[0]
        if isinstance(first, dict):
            selected = first
    if selected is None:
        return response

    if singular_key in selected and isinstance(selected[singular_key], dict):
        return {singular_key: selected[singular_key]}
    return {singular_key: selected}


def unwrap_ipv4(item: dict[str, Any], wrapped_key: str) -> dict[str, Any] | None:
    """Return IPv4 payload from either wrapped or direct item."""
    if wrapped_key in item and isinstance(item[wrapped_key], dict):
        inner = item[wrapped_key]
        ipv4 = inner.get("ipv4")
        if isinstance(ipv4, dict):
            return ipv4
        return inner
    ipv4 = item.get("ipv4")
    if isinstance(ipv4, dict):
        return ipv4
    return None

File: packages/typescript/src/resources/normalize.ts
/**
 * Shared response normalization (firmware envelope variants).
 */

export function normalizeGetFromPlural(
  response: Record<string, unknown>,
  options: {
    pluralKey: string;
    singularKey: string;
    predicate?: (item: Record<string, unknown>) => boolean;
  }
): Record<string, unknown> {
  const { pluralKey, singularKey, predicate } = options;
  const items = response[pluralKey];
  if (!Array.isArray(items) || items.length === 0) {
    return response;
  }

  let selected: Record<string, unknown> | null = null;
  if (predicate) {
    for (const item of items) {
      if (item && typeof item === "object" && predicate(item as Record<string, unknown>)) {
        selected = item as Record<string, unknown>;
        break;
      }
    }
  }
  if (selected === null) {
    const first = items[0];
    if (first && typeof first === "object") {
      selected = first as Record<string, unknown>;
    }
  }
  if (selected === null) {
    return response;
  }

  const inner = selected[singularKey];
  if (singularKey in selected && typeof inner === "object" && inner !== null) {
    return { [singularKey]: inner };
  }
  return { [singularKey]: selected };
}

export function unwrapIpv4(
  item: Record<string, unknown>,
  wrappedKey: string
): Record<string, unknown> | null {
  if (wrappedKey in item && typeof item[wrappedKey] === "object" && item[wrappedKey] !== null) {
    const inner = item[wrappedKey] as Record<string, unknown>;
    const ipv4 = inner["ipv4"];
    if (typeof ipv4 === "object" && ipv4 !== null) {
      return ipv4 as Record<string, unknown>;
    }
    return inner;
  }
  const ipv4 = item["ipv4"];
  if (typeof ipv4 === "object" && ipv4 !== null) {
    return ipv4 as Record<string, unknown>;
  }
  return null;
}
